roofit/roofitcore/src/RooGradMinimizerFcn.cxx

```diff
@@ -0,0 +1,238 @@
+            ooccoutW(static_cast<RooAbsArg *>(nullptr), Eval) << var->GetName() << "=" << var->getVal();
+         }
+         delete iter;
+         ooccoutW(static_cast<RooAbsArg *>(nullptr), Eval) << std::endl;
+
+         RooAbsReal::printEvalErrors(ooccoutW(static_cast<RooAbsArg *>(nullptr), Eval), _printEvalErrors);
+         ooccoutW(static_cast<RooAbsArg *>(nullptr), Eval) << std::endl;
+      }
+
+      if (_doEvalErrorWall) {
+         fvalue = _maxFCN + 1;
+      }
+
+      RooAbsReal::clearEvalErrorLog();
+      _numBadNLL++;
+   } else if (fvalue > _maxFCN) {
+      _maxFCN = fvalue;
+   }
+
+   // Optional logging
+   if (_verbose) {
+      std::cout << "\nprevFCN" << (_funct->isOffsetting() ? "-offset" : "") << " = " << std::setprecision(10) << fvalue
+                << std::setprecision(4) << "  ";
+      std::cout.flush();
+   }
+
+   _evalCounter++;
+   //#ifndef NDEBUG
+   //  std::cout << "RooGradMinimizerFcn " << this << " evaluations (in DoEval): " << _evalCounter <<
+   //  std::endl;
+   //#endif
+   return fvalue;
+}
+
+void RooGradMinimizerFcn::resetHasBeenCalculatedFlags() const
+{
+   for (auto it = has_been_calculated.begin(); it != has_been_calculated.end(); ++it) {
+      *it = false;
+   }
+   none_have_been_calculated = true;
+}
+
+bool RooGradMinimizerFcn::syncParameter(double x, std::size_t ix) const
+{
+   bool parameter_has_changed = (_grad_params[ix] != x);
+
+   if (parameter_has_changed) {
+      _grad_params[ix] = x;
+      // Set the parameter values for this iteration
+      // TODO: this is already done in DoEval as well; find efficient way to do only once
+      SetPdfParamVal(ix, x);
+
+      if (!none_have_been_calculated) {
+         resetHasBeenCalculatedFlags();
+      }
+   }
+
+   return parameter_has_changed;
+}
+
+bool RooGradMinimizerFcn::syncParameters(const double *x) const
+{
+   bool has_been_synced = false;
+
+   for (std::size_t ix = 0; ix < NDim(); ++ix) {
+      bool parameter_has_changed = (_grad_params[ix] != x[ix]);
+
+      if (parameter_has_changed) {
+         _grad_params[ix] = x[ix];
+         // Set the parameter values for this iteration
+         // TODO: this is already done in DoEval as well; find efficient way to do only once
+         SetPdfParamVal(ix, x[ix]);
+      }
+
+      has_been_synced |= parameter_has_changed;
+   }
+
+   if (has_been_synced) {
+      resetHasBeenCalculatedFlags();
+   }
+
+   return has_been_synced;
+}
+
+void RooGradMinimizerFcn::runDerivator(unsigned int i_component) const
+{
+   // check whether the derivative was already calculated for this set of parameters
+   if (!has_been_calculated[i_component]) {
+      // Calculate the derivative etc for these parameters
+      _grad[i_component] =
+         _gradf.partial_derivative(_context->getMultiGenFcn(), _grad_params.data(),
+                                   _context->fitter()->Config().ParamsSettings(), i_component, _grad[i_component]);
```

Comment:
```suggestion
         _gradf.PartialDerivative(_context->getMultiGenFcn(), _grad_params.data(),
                                  _context->fitter()->Config().ParamsSettings(), i_component, _grad[i_component]);
```

Reply:
Ugh, man, sorry! Lemme fix the compile errors first (compilation still ongoing on my laptop, it has to recompile everything all the time due to switching branches a lot) before you waste any more time reviewing trivial stuff :)